src/Handlers/PaymentResponseHandler.php

```diff
@@ -116,60 +118,65 @@ public function __construct(
     }
 
     /**
-     * @param array $response
+     * @param PaymentResponse|PaymentDetailsResponse $response
      * @param OrderTransactionEntity $orderTransaction
+     * @param bool $upsertResponse
      * @return PaymentResponseHandlerResult
      */
     public function handlePaymentResponse(
-        array $response,
+        $response,
         OrderTransactionEntity $orderTransaction,
         bool $upsertResponse = true
     ): PaymentResponseHandlerResult {
-        $paymentResponseHandlerResult = new PaymentResponseHandlerResult();
-        // Retrieve result code from response array
-        $resultCode = $response['resultCode'];
-        if (array_key_exists('refusalReason', $response)) {
-            $paymentResponseHandlerResult->setRefusalReason($response['refusalReason']);
-        }
 
-        if (array_key_exists('refusalReasonCode', $response)) {
-            $paymentResponseHandlerResult->setRefusalReasonCode($response['refusalReasonCode']);
+        if (!($response instanceof PaymentResponse) && !($response instanceof PaymentDetailsResponse)) {
+            throw new \InvalidArgumentException('Invalid $paymentDetailsResponse type.');
         }
 
-        $paymentResponseHandlerResult->setResultCode($resultCode);
+        // Retrieve result code from response array
+        $resultCode = $response->getResultCode();
 
-        // Retrieve PSP reference from response array if available
-        if (!empty($response[self::PSP_REFERENCE])) {
-            $paymentResponseHandlerResult->setPspReference($response[self::PSP_REFERENCE]);
-        }
+        $paymentResponseHandlerResult = new PaymentResponseHandlerResult();
+        $paymentResponseHandlerResult->setRefusalReason($response->getRefusalReason());
+        $paymentResponseHandlerResult->setRefusalReasonCode($response->getRefusalReasonCode());
+        $paymentResponseHandlerResult->setResultCode($resultCode);
 
-        // Set action in result object if available
-        if (!empty($response[self::ACTION])) {
-            $paymentResponseHandlerResult->setAction($response[self::ACTION]);
+        // when this method is called from PaymentDetailsResponse, $response would not have the action part
+        if (method_exists($response, 'getAction') && !empty($response->getAction())) {
+            if (is_array($response->getAction())) {
+                // this is the case when handlePaymentResponse is being called
+                // in AbstractPaymentMethodHandler with $giftcardPaymentResponse
+                // in this use-case, $giftcardPaymentResponse->action is an array and not an object
+                $paymentResponseHandlerResult->setAction($response->getAction());
+            } else {
+                $paymentResponseHandlerResult->setAction($response->getAction()->jsonSerialize());
```

Comment:
Using `toArray()` instead of `jsonSerialize()` might be better since `setAction` expects an array.